### PR TITLE
initial integration of Andrew Apted voxel code

### PR DESCRIPTION
**Pull request by rfomin**
Need more work for proper clipping. @fabiangreffrath Should we continue?

Author's repository: https://gitlab.com/andwj/voxel-rend

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
> Need more work for proper clipping.

👀 Wow, that was fast. I have just seen the thread, but am still trapped at day job.


> @fabiangreffrath Should we continue?

Yes, of course! 😁

**Comment by fabiangreffrath** [comment]
I don't understand `ISO C90 forbids mixed declarations and code` errors in CI. Don't we say C99 here?
https://github.com/fabiangreffrath/woof/blob/400a42e6028104dd7f8069fb2f77b8a5d5d02fde/CMakeLists.txt#L19

**Comment by fabiangreffrath** [comment]
> Edit: We explicitly set `-Wdeclaration-after-statement`. Maybe remove that line for the time working on this PR?

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,8 +16,6 @@ project("Woof"
     VERSION 12.0.0
     LANGUAGES C)
 
-set(CMAKE_C_STANDARD 99)
```
Comment (fabiangreffrath) [inline CMakeLists.txt:19]: Could you move back to where it was? It was actually doing its job, we just set this extra warning.

Comment (rfomin) [reply CMakeLists.txt:19]: Does it work? I think we should use https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#manual:cmake-compile-features(7)

**Comment by rfomin** [comment]
Currently crashing in nuts.wad. This is of course the absolute extreme, but I've tested other big maps and performance isn't that bad in the release version (debug is very slow).

File: CMakeLists.txt (hunk 1 above)
Comment (fabiangreffrath) [reply CMakeLists.txt:19]: Yes, it does. It sets the default value for the `C_STANDARD` feature, which provides a fallback mechanism for compilers which doe not support specific feature sets:

https://cmake.org/cmake/help/latest/prop_tgt/C_STANDARD.html

In our case, it indeed adds e.g. the `-std=gnu99` flag to the GCC call.

Comment (fabiangreffrath) [reply CMakeLists.txt:19]: We should probably even `set(CMAKE_C_STANDARD_REQUIRED ON)`. 🤔

**Comment by fabiangreffrath** [comment]
> Currently crashing in nuts.wad. This is of course the absolute extreme, but I've tested other big maps and performance isn't that bad in the release version (debug is very slow).

Could you figure out where it crashes? Is this an OOM issue that we could catch?

**Comment by fabiangreffrath** [comment]
> Currently crashing in nuts.wad. This is of course the absolute extreme, but I've tested other big maps and performance isn't that bad in the release version (debug is very slow).

Does it help if you cap the number of actually drawn sprites/voxels at a certain value?

```diff
--- a/src/r_things.c
+++ b/src/r_things.c
@@ -1165,7 +1165,7 @@ void R_DrawMasked(void)

   // draw all vissprites back to front

-  for (i = num_vissprite ;--i>=0; )
+  for (i = MIN(num_vissprite, 4096) ;--i>=0; )
   {
     vissprite_t* spr = vissprite_ptrs[i];

```

**Comment by rfomin** [comment]
> Does it help if you cap the number of actually drawn sprites/voxels at a certain value?

It already does a smart thing, and does not draw voxels far from the player. Some use after free issue, ASan report:
<details>

```
=================================================================
==4216==ERROR: AddressSanitizer: heap-use-after-free on address 0x11d674c35ae8 at pc 0x7ff84367face bp 0x007bd495bdf0 sp 0x007bd495bdf0
READ of size 8 at 0x11d674c35ae8 thread T0
    #0 0x7ff84367facd in VX_DrawVoxel C:\develop\woof\src\r_voxel.c:875
    #1 0x7ff84367a6d9 in R_DrawSprite C:\develop\woof\src\r_things.c:1113
    #2 0x7ff843673c5f in R_DrawMasked C:\develop\woof\src\r_things.c:1200
    #3 0x7ff8436674a2 in R_RenderPlayerView C:\develop\woof\src\r_main.c:840
    #4 0x7ff84351cccb in D_Display C:\develop\woof\src\d_main.c:285
    #5 0x7ff843525243 in D_DoomMain C:\develop\woof\src\d_main.c:2843
    #6 0x7ff843556d7b in Woof_Main C:\develop\woof\src\i_main.c:58
    #7 0x7ff7d27e278a in SDL_main C:\develop\woof\win32\win_launcher.c:21
    #8 0x7ff7d27e89aa in main_getcmdline C:\develop\vcpkg\buildtrees\sdl2\src\ase-2.28.3-38b1860324.clean\src\main\windows\SDL_windows_main.c:80
    #9 0x7ff7d27e8a52 in main C:\develop\vcpkg\buildtrees\sdl2\src\ase-2.28.3-38b1860324.clean\src\main\windows\SDL_windows_main.c:94
    #10 0x7ff7d27e3538 in invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
    #11 0x7ff7d27e341d in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #12 0x7ff7d27e32dd in __scrt_common_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:330
    #13 0x7ff7d27e35cd in mainCRTStartup D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:16
    #14 0x7ff877607343 in BaseThreadInitThunk+0x13 (C:\Windows\System32\KERNEL32.DLL+0x180017343)
    #15 0x7ff8780a26b0 in RtlUserThreadStart+0x20 (C:\Windows\SYSTEM32\ntdll.dll+0x1800526b0)

0x11d674c35ae8 is located 3560 bytes inside of 4136-byte region [0x11d674c34d00,0x11d674c35d28)
freed by thread T0 here:
    #0 0x7ff8429d33b4 in _asan_wrap_GlobalSize+0x5036b (c:\develop\woof\build_asan\src\clang_rt.asan_dbg_dynamic-x86_64.dll+0x1800533b4)
    #1 0x7ff8436a92f7 in Z_Free C:\develop\woof\src\z_zone.c:115
    #2 0x7ff8436a9c5e in Z_Realloc C:\develop\woof\src\z_zone.c:188
    #3 0x7ff84367c6fa in R_NewVisVoxel C:\develop\woof\src\r_voxel.c:308
    #4 0x7ff84367d880 in VX_ProjectVoxel C:\develop\woof\src\r_voxel.c:530
    #5 0x7ff8436753c8 in R_ProjectSprite C:\develop\woof\src\r_things.c:470
    #6 0x7ff843673011 in R_AddSprites C:\develop\woof\src\r_things.c:702
    #7 0x7ff84365b288 in R_Subsector C:\develop\woof\src\r_bsp.c:644
    #8 0x7ff84365b4de in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:681
    #9 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #10 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #11 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #12 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #13 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #14 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #15 0x7ff843667470 in R_RenderPlayerView C:\develop\woof\src\r_main.c:822
    #16 0x7ff84351cccb in D_Display C:\develop\woof\src\d_main.c:285
    #17 0x7ff843525243 in D_DoomMain C:\develop\woof\src\d_main.c:2843
    #18 0x7ff843556d7b in Woof_Main C:\develop\woof\src\i_main.c:58
    #19 0x7ff7d27e278a in SDL_main C:\develop\woof\win32\win_launcher.c:21
    #20 0x7ff7d27e89aa in main_getcmdline C:\develop\vcpkg\buildtrees\sdl2\src\ase-2.28.3-38b1860324.clean\src\main\windows\SDL_windows_main.c:80
    #21 0x7ff7d27e8a52 in main C:\develop\vcpkg\buildtrees\sdl2\src\ase-2.28.3-38b1860324.clean\src\main\windows\SDL_windows_main.c:94
    #22 0x7ff7d27e3538 in invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
    #23 0x7ff7d27e341d in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #24 0x7ff7d27e32dd in __scrt_common_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:330
    #25 0x7ff7d27e35cd in mainCRTStartup D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:16
    #26 0x7ff877607343 in BaseThreadInitThunk+0x13 (C:\Windows\System32\KERNEL32.DLL+0x180017343)
    #27 0x7ff8780a26b0 in RtlUserThreadStart+0x20 (C:\Windows\SYSTEM32\ntdll.dll+0x1800526b0)

previously allocated by thread T0 here:
    #0 0x7ff8429d3554 in _asan_wrap_GlobalSize+0x5050b (c:\develop\woof\build_asan\src\clang_rt.asan_dbg_dynamic-x86_64.dll+0x180053554)
    #1 0x7ff8436a8c31 in Z_Malloc C:\develop\woof\src\z_zone.c:62
    #2 0x7ff8436a9b42 in Z_Realloc C:\develop\woof\src\z_zone.c:183
    #3 0x7ff84367c6fa in R_NewVisVoxel C:\develop\woof\src\r_voxel.c:308
    #4 0x7ff84367d880 in VX_ProjectVoxel C:\develop\woof\src\r_voxel.c:530
    #5 0x7ff8436753c8 in R_ProjectSprite C:\develop\woof\src\r_things.c:470
    #6 0x7ff843673011 in R_AddSprites C:\develop\woof\src\r_things.c:702
    #7 0x7ff84365b288 in R_Subsector C:\develop\woof\src\r_bsp.c:644
    #8 0x7ff84365b4de in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:681
    #9 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #10 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #11 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #12 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #13 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #14 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #15 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #16 0x7ff84365b41f in R_RenderBSPNode C:\develop\woof\src\r_bsp.c:672
    #17 0x7ff843667470 in R_RenderPlayerView C:\develop\woof\src\r_main.c:822
    #18 0x7ff84351cccb in D_Display C:\develop\woof\src\d_main.c:285
    #19 0x7ff843525243 in D_DoomMain C:\develop\woof\src\d_main.c:2843
    #20 0x7ff843556d7b in Woof_Main C:\develop\woof\src\i_main.c:58
    #21 0x7ff7d27e278a in SDL_main C:\develop\woof\win32\win_launcher.c:21
    #22 0x7ff7d27e89aa in main_getcmdline C:\develop\vcpkg\buildtrees\sdl2\src\ase-2.28.3-38b1860324.clean\src\main\windows\SDL_windows_main.c:80
    #23 0x7ff7d27e8a52 in main C:\develop\vcpkg\buildtrees\sdl2\src\ase-2.28.3-38b1860324.clean\src\main\windows\SDL_windows_main.c:94
    #24 0x7ff7d27e3538 in invoke_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
    #25 0x7ff7d27e341d in __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #26 0x7ff7d27e32dd in __scrt_common_main D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:330
    #27 0x7ff7d27e35cd in mainCRTStartup D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp:16
    #28 0x7ff877607343 in BaseThreadInitThunk+0x13 (C:\Windows\System32\KERNEL32.DLL+0x180017343)

SUMMARY: AddressSanitizer: heap-use-after-free C:\develop\woof\src\r_voxel.c:875 in VX_DrawVoxel
Shadow bytes around the buggy address:
  0x03cf43286b00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x03cf43286b10: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x03cf43286b20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x03cf43286b30: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x03cf43286b40: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x03cf43286b50: fd fd fd fd fd fd fd fd fd fd fd fd fd[fd]fd fd
  0x03cf43286b60: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x03cf43286b70: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x03cf43286b80: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x03cf43286b90: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x03cf43286ba0: fd fd fd fd fd fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==4216==ABORTING
```
</details>

**Comment by rfomin** [comment]
Found the problem, it's my sloppy coding. Need to think a bit 😄

File: src/r_voxel.c [hunk 2]
```diff
@@ -0,0 +1,962 @@
...
+static boolean VX_Load (int spr, int frame)
+{
+	char frame_ch = 'A' + frame;
+
+	if (frame_ch == '\\')
+		frame_ch = '^';
+
+	char filename[256];
+
+	M_snprintf (filename, sizeof(filename), "voxels/%s%c.kvx", sprnames[spr], frame_ch);
```
Comment (fabiangreffrath) [inline src/r_voxel.c:238]: How are we going to handle the data loading? Autoload directory? Command line parameter?

Comment (MrAlaux) [reply src/r_voxel.c:238]: If you don't mind me chiming in, I was thinking about a `VX_` namespace in WADs, to make this more viable for modders.

Comment (rfomin) [reply src/r_voxel.c:238]: I think we should add support for the `.kvx` extension and load it from `autoload` or `.zip` files. Maybe we should also support the `voxels` directory.

File: src/r_voxel.c [hunk 3]
```diff
@@ -0,0 +1,962 @@
...
+	fixed_t BL_y = TL_y - (ys * c);
+
+	fixed_t TR_x = TL_x + (xs * c);
+	fixed_t TR_y = TL_y + (xs * s);
+
+	fixed_t BR_x = BL_x + (xs * c);
+	fixed_t BR_y = BL_y + (xs * s);
+
+	// is it outside the view frustum?
+	if (! VX_CheckFrustum (TL_x, TL_y, BL_x, BL_y, BR_x, BR_y, TR_x, TR_y))
```
Comment (rfomin) [inline src/r_voxel.c:482]: I fixed crashes, everything works except the widescreen clipping. The `VX_CheckFrustrum()` function works only for 4:3 aspect ratio.

File: src/r_voxel.c (hunk 2 above)
Comment (rfomin) [reply src/r_voxel.c:238]: > If you don't mind me chiming in, I was thinking about a `VX_` namespace in WADs, to make this more viable for modders.

That's a good idea too. I don't think anyone supports voxels in WADs, right?

Comment (MrAlaux) [reply src/r_voxel.c:238]: > That's a good idea too. I don't think anyone supports voxels in WADs, right?

I mean, I don't know if other ports would ignore stuff between markers they don't recognize. Optimally, they should load e.g. a `BON1A0` between `S_` (or `SS_`) markers but not a `BON1A0` between something like these `VX_` markers I'm proposing.

On that note, with voxels we don't have to worry about rotations, so I think another possibility would be to include voxels in WADs by giving them a prefix: `VOX[sprite]`, e.g. `VOXBON1A`. That shouldn't conflict with normal sprites and probably wouldn't need markers.

**Comment by fabiangreffrath** [comment]
Working nice so far! I even get two-digit FPS on nuts.wad, though the Demon corpse sprites seem to be performance hugs for some reason.

One thing that bugs me, though, is how static sprites rotate. Have a look, for example, how the rocket launcher seems to rotate with my view in the following clip:

[Bildschirmaufzeichnung vom 2023-11-08, 21-49-56.webm](https://github.com/fabiangreffrath/woof/assets/6449237/e83a87d9-1ae3-4201-9078-7777c78f89a9)

I got the clipping "fixed" in widescreen mode by using `clipangle/2` in `VX_CheckFrustum()`, but that's obviously a hack.

**Comment by fabiangreffrath** [comment]
Another thing, this overrides sprites replacements from PWADs (e.g. the dark imps from Eternal Doom).

And why are the "pixels" on the Caco's horns shaped like pillows?

grafik

File: src/r_voxel.c (hunk 2 above)
Comment (kraflab) [reply src/r_voxel.c:238]: I believe ports like pr+ and dsda-doom _would_ load stuff between markers that are unknown.

Comment (MrAlaux) [reply src/r_voxel.c:238]: Sounds like that'd require a cross-port effort, which I'd personally rather avoid. So, using the exact names of the sprites we're providing voxels for is a no-go.

Perhaps we could combine both of the approaches I proposed: use a `VX_` namespace to tell our voxel-supporting port that the lumps in that namespace are indeed voxels, and also use a prefix or suffix to not confuse non-voxel-supporting ports into trying to load voxels as sprites.

Maybe the prefix/suffix part is unnecessary; I'm also not sure if ports try to load sprites outside of the `S_` namespace.

Comment (rfomin) [reply src/r_voxel.c]: The *.kvm files have no headers, so identifying them is rather tedious: https://github.com/ZDoom/gzdoom/blob/d5e9783324936b50eb4b7b8ed646894a30e226b2/src/common/models/voxels.cpp#L168
So I think load from WADs and *.lmp is not a good idea.

With the last commit we can load *.kvm from *.zip and *.pk3, -file parameter, autoload directory and `voxels` directory. `cheello_voxels_v2_1.pk3` just works.

**Comment by rfomin** [comment]
> I even get two-digit FPS on nuts.wad, though the Demon corpse sprites seem to be performance hugs for some reason.

I find the piles of corpses slow in general. Maybe we should replace them dynamically with sprites. There is `VX_NEAR_RADIUS` and `VX_MAX_DIST`, perhaps we should add the amount.

**Comment by rfomin** [comment]
> I got the clipping "fixed" in widescreen mode by using `clipangle/2` in `VX_CheckFrustum()`, but that's obviously a hack.

Well, that hack works 😄 Any other ideas?

**Comment by fabiangreffrath** [comment]
> I find the piles of corpses slow in general. Maybe we should replace them dynamically with sprites. There is `VX_NEAR_RADIUS` and `VX_MAX_DIST`, perhaps we should add the amount.

I understand that voxel which are further away than `VX_MAX_DIST` are rendered as sprites. But what is the purpose of `VX_NEAR_RADIUS`?

**Comment by MrAlaux** [comment]
Found a bug -- Voxels are shifted as the Screen Size decreases:

<details>
  <summary>Screenshots</summary>

woof0000

woof0001
</details>

**Comment by MrAlaux** [comment]
On the topic of `VX_MAX_DIST`, I think it could use some fine-tuning to increase performance: IMO, it's higher than necessary in low resolution, and even in high resolution it could probably be decreased a bit.

Maybe try making it `((x * FRACUNIT) << hires)`, where `x` is somewhere between 800 and 1000.

**Comment by fabiangreffrath** [comment]
Got some savegames with strange artifacts (Doom 1, E3M1):

grafik

grafik


[e3m1.zip](https://github.com/fabiangreffrath/woof/files/13308511/e3m1.zip)

**Comment by rfomin** [comment]
> But what is the purpose of `VX_NEAR_RADIUS`?

Actually, I'm not sure. I misunderstood the code.

> Found a bug -- Voxels are shifted as the Screen Size decreases. Maybe the hack that fixes voxels in widescreen also fixes this?

Yes, another problem that seems to be present in the original implementation as well. Changes in `VX_CheckFrustum()` will not help here.

> On the topic of `VX_MAX_DIST`, I think it could use some fine-tuning to increase performance: IMO, it's higher than necessary in low resolution, and even in high resolution it could probably be decreased a bit.

I was thinking about dynamic `VX_MAX_DIST` change, depend on frame rendering time or quantity of voxel models in frustum.

**Comment by rfomin** [comment]
> Got some savegames with strange artifacts (Doom 1, E3M1):

This is related to the `clipangle / 2` hack. Looks like we have to fix it the hard way  😉

File: src/r_voxel.c (hunk 2 above)
Comment (fabiangreffrath) [reply src/r_voxel.c]: > With the last commit we can load *.kvm from *.zip and *.pk3, -file parameter, autoload directory and `voxels` directory. `cheello_voxels_v2_1.pk3` just works.

This is the best solution, awesome! 👍

File: src/r_voxel.c (hunk 3 above)
Comment (ceski-1) [reply src/r_voxel.c:523]: This is one way to fix widescreen:

<details><summary>clip angle</summary>

```diff
diff --git a/src/r_main.c b/src/r_main.c
index b56e486..16dccb1 100644
--- a/src/r_main.c
+++ b/src/r_main.c
@@ -489,6 +489,7 @@ void R_ExecuteSetViewSize (void)
   R_InitBuffer(scaledviewwidth, scaledviewheight);       // killough 11/98
         
   R_InitTextureMapping();
+  VX_UpdateClipAngle();
     
   // psprite scales
   pspritescale = FixedDiv(viewwidth_nonwide, ORIGWIDTH);       // killough 11/98
```

```diff
diff --git a/src/r_voxel.c b/src/r_voxel.c
index bae6d05..74ce9dd 100644
--- a/src/r_voxel.c
+++ b/src/r_voxel.c
@@ -382,6 +382,15 @@ static int VX_RotateModeForThing (mobj_t * thing)
 }
 
 
+#define VX_PI 3.14159265358979323846
+static angle_t vx_clipangle;
+
+void VX_UpdateClipAngle(void)
+{
+	const double fovrad = 2 * atan(SCREENWIDTH / (1.2 * SCREENHEIGHT) * 3 / 4);
+	vx_clipangle = fovrad / VX_PI * ANG180 - ANG90;
+}
+
 static boolean VX_CheckFrustum (fixed_t x1, fixed_t y1, fixed_t x2, fixed_t y2,
                          fixed_t x3, fixed_t y3, fixed_t x4, fixed_t y4)
 {
@@ -389,8 +398,8 @@ static boolean VX_CheckFrustum (fixed_t x1, fixed_t y1, fixed_t x2, fixed_t y2,
 	if (y1 < VX_MINZ && y2 < VX_MINZ && y3 < VX_MINZ && y4 < VX_MINZ)
 		return false;
 
-	fixed_t c = finecosine[clipangle >> ANGLETOFINESHIFT];
-	fixed_t s = finesine  [clipangle >> ANGLETOFINESHIFT];
+	fixed_t c = finecosine[vx_clipangle >> ANGLETOFINESHIFT];
+	fixed_t s = finesine  [vx_clipangle >> ANGLETOFINESHIFT];
 
 	// note the funky shifts here, to counter-balance opposite
 	// shifts in the R_PointOnSide() code.  Sigh.
```

```diff
diff --git a/src/r_voxel.h b/src/r_voxel.h
index 8ec92d1..e6828dc 100644
--- a/src/r_voxel.h
+++ b/src/r_voxel.h
@@ -27,4 +27,6 @@ boolean VX_ProjectVoxel (mobj_t * thing);
 
 void VX_DrawVoxel (vissprite_t * vis);
 
+void VX_UpdateClipAngle(void);
+
 #endif  /* __R_VOXEL__ */
```

</details>

Comment (rfomin) [reply src/r_voxel.c:523]: Thanks for idea, this seems to work:

<details>

```diff
diff --git a/src/i_video.c b/src/i_video.c
index 89263d91..058ef4f8 100644
--- a/src/i_video.c
+++ b/src/i_video.c
@@ -45,6 +45,7 @@
 int SCREENWIDTH, SCREENHEIGHT;
 int NONWIDEWIDTH; // [crispy] non-widescreen SCREENWIDTH
 int WIDESCREENDELTA; // [crispy] horizontal widescreen offset
+int FOV;
 
 boolean use_vsync;  // killough 2/8/98: controls whether vsync is called
 int hires, default_hires;      // killough 11/98
@@ -958,6 +959,13 @@ void I_GetScreenDimensions(void)
     }
 
     WIDESCREENDELTA = (SCREENWIDTH - NONWIDEWIDTH) / 2;
+
+    // The general equation is as follows:
+    // y = mx + b -> fov = (75/2) * ratio + 40
+    // This gives 90 for 4:3, 100 for 16:10, and 106 for 16:9.
+
+    double ratio = (double)SCREENWIDTH / (double)ah;
+    FOV = (int)((75.0/2.0) * ratio + 40.0);
 }
 
 static void CreateUpscaledTexture(boolean force)
diff --git a/src/i_video.h b/src/i_video.h
index 6afa4399..dbf67de1 100644
--- a/src/i_video.h
+++ b/src/i_video.h
@@ -28,6 +28,7 @@ extern int SCREENWIDTH;
 extern int SCREENHEIGHT;
 extern int NONWIDEWIDTH; // [crispy] non-widescreen SCREENWIDTH
 extern int WIDESCREENDELTA; // [crispy] horizontal widescreen offset
+extern int FOV;
 void I_GetScreenDimensions (void); // [crispy] re-calculate WIDESCREENDELTA
 
 enum
diff --git a/src/r_main.c b/src/r_main.c
index b56e486d..d20789d1 100644
--- a/src/r_main.c
+++ b/src/r_main.c
@@ -57,6 +57,7 @@ fixed_t  viewheightfrac; // [FG] sprite clipping optimizations
 //
 
 angle_t clipangle;
+angle_t vx_clipangle;
 
 // The viewangletox[viewangle + FINEANGLES/4] lookup
 // maps the visible view angles to screen X coordinates,
@@ -305,6 +306,8 @@ static void R_InitTextureMapping (void)
         viewangletox[i] = viewwidth;
         
   clipangle = xtoviewangle[0];
+
+  vx_clipangle = clipangle - (FOV - 90) * ANG1;
 }
 
 //
diff --git a/src/r_state.h b/src/r_state.h
index e6668610..69ca7abe 100644
--- a/src/r_state.h
+++ b/src/r_state.h
@@ -96,6 +96,7 @@ extern fixed_t          viewz;
 extern angle_t          viewangle;
 extern player_t         *viewplayer;
 extern angle_t          clipangle;
+extern angle_t          vx_clipangle;
 extern int              viewangletox[FINEANGLES/2];
 extern angle_t          xtoviewangle[MAX_SCREENWIDTH+1];  // killough 2/8/98
 extern fixed_t          rw_distance;
diff --git a/src/r_voxel.c b/src/r_voxel.c
index bae6d052..ff05dee5 100644
--- a/src/r_voxel.c
+++ b/src/r_voxel.c
@@ -389,8 +389,8 @@ static boolean VX_CheckFrustum (fixed_t x1, fixed_t y1, fixed_t x2, fixed_t y2,
 	if (y1 < VX_MINZ && y2 < VX_MINZ && y3 < VX_MINZ && y4 < VX_MINZ)
 		return false;
 
-	fixed_t c = finecosine[clipangle >> ANGLETOFINESHIFT];
-	fixed_t s = finesine  [clipangle >> ANGLETOFINESHIFT];
+	fixed_t c = finecosine[vx_clipangle >> ANGLETOFINESHIFT];
+	fixed_t s = finesine  [vx_clipangle >> ANGLETOFINESHIFT];
 
 	// note the funky shifts here, to counter-balance opposite
 	// shifts in the R_PointOnSide() code.  Sigh.
```
</details>

I think we need to add a custom FOV anyway.

Comment (ceski-1) [reply src/r_voxel.c:523]: Unfortunately it seems to flicker: [allmonsters.zip](https://github.com/fabiangreffrath/woof/files/13316051/allmonsters.zip)
```
woof -file allmonsters.wad -playdemo allmonsters
```

Comment (rfomin) [reply src/r_voxel.c:523]: > Unfortunately it seems to flicker: [allmonsters.zip](https://github.com/fabiangreffrath/woof/files/13316051/allmonsters.zip)

It flickers also with normal `clipangle` when widescreen mode is off. I think there are problems with very large models.

Comment (ceski-1) [reply src/r_voxel.c:523]: Okay, it looks good to me other than that.

Comment (fabiangreffrath) [reply src/r_voxel.c:523]: Not sure if he has an account on github, but maybe we should ask Andrew Apted for his input?

Comment (rfomin) [reply src/r_voxel.c:523]: Flickering disappears in 200p mode. I missed `hires` somewhere.

Comment (rfomin) [reply src/r_voxel.c:543]: Fixed `hires` flickering: https://github.com/fabiangreffrath/woof/pull/1249/commits/89e71c83a6531310109cd73dc08ba4ddcd9714e3 [allmonsters.zip](https://github.com/fabiangreffrath/woof/files/13316051/allmonsters.zip) demo works

Comment (fabiangreffrath) [reply src/r_voxel.c:543]: Awesome, thank you! How about the "view size" issue?

Comment (rfomin) [reply src/r_voxel.c:543]: > How about the "view size" issue?

Don't know yet. I guess we need to use `viewwindowx` and `viewwindowy` offsets.

Comment (ceski-1) [reply src/r_voxel.c:543]: > Fixed `hires` flickering: [89e71c8](https://github.com/fabiangreffrath/woof/commit/89e71c83a6531310109cd73dc08ba4ddcd9714e3) [allmonsters.zip](https://github.com/fabiangreffrath/woof/files/13316051/allmonsters.zip) demo works

I did a quick test and it looks good.

Comment (rfomin) [reply src/r_voxel.c:543]: > How about the "view size" issue?

Fixed in https://github.com/fabiangreffrath/woof/pull/1249/commits/258742b646f3a6b88cb02e67b6c4008cadcaf062

Comment (fabiangreffrath) [reply src/r_voxel.c:543]: Awesome!

File: src/r_main.c [hunk 4]
```diff
@@ -818,7 +823,9 @@ void R_RenderPlayerView (player_t* player)
 
   // The head node is the last node output.
   R_RenderBSPNode (numnodes-1);
-    
+
```
Comment (fabiangreffrath) [inline src/r_main.c:826]: What does this do if no voxels are available?

Comment (fabiangreffrath) [reply src/r_main.c:826]: > What does this do if no voxels are available?

What I mean is, I am afraid that this affects regular sprite rendering.

Comment (rfomin) [reply src/r_main.c:826]: If no model exists for the given sprite? Then `VX_ProjectVoxel()` exits, and `R_ProjectSprite()` continues to draw the standard sprite.

`VX_NearbySprites()` handles the special case of models disappearing in small sectors that have gone out of view. I haven't tested it, I'm not sure how often it's needed.

> What I mean is, I am afraid that this affects regular sprite rendering.

As far as I know, it doesn't affect it in any way. The code looks good.

Comment (fabiangreffrath) [reply src/r_main.c:826]: Thank you!

**Review by fabiangreffrath — APPROVED** [review]

**Pull request merged**